File: src/main/java/com/sleepkqq/sololeveling/service/auth/TgHashService.java
package com.sleepkqq.sololeveling.service.auth;

import static java.nio.charset.StandardCharsets.UTF_8;

import com.sleepkqq.sololeveling.view.auth.TgAuthData;
import java.net.URLDecoder;
import java.security.GeneralSecurityException;
import java.util.Map;
import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import lombok.extern.slf4j.Slf4j;
import one.util.streamex.EntryStream;
import one.util.streamex.StreamEx;
import org.bouncycastle.util.encoders.Hex;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

@Slf4j
@Service
public class TgHashService {

  private static final String TG_SECRET_KEY = "WebAppData";

  private static final String HMAC_SHA256 = "HmacSHA256";

  private static final String HASH_FIELD = "hash";

  @Value("${telegram.bot.token}")
  private String tgBotToken;

  public boolean checkHash(TgAuthData tgAuthData) {
    var parsedQueryString = parseQueryString(tgAuthData.initData());
    try {
      return validateTelegramAuth(parsedQueryString, parsedQueryString.get(HASH_FIELD));
    } catch (GeneralSecurityException e) {
      log.error("Hash check algorithm error", e);
      return false;
    }
  }

  private boolean validateTelegramAuth(
      Map<String, String> paramMap,
      String receivedHash
  ) throws GeneralSecurityException {
    var dataString = EntryStream.of(paramMap)
        .filterKeys(k -> !HASH_FIELD.equals(k))
        .sorted(Map.Entry.comparingByKey())
        .mapKeyValue((k, v) -> k + "=" + v)
        .joining("\n");

    var sha256HMAC = Mac.getInstance(HMAC_SHA256);
    var secretKeySpec = new SecretKeySpec(getSecretHashByInitData(), HMAC_SHA256);
    sha256HMAC.init(secretKeySpec);

    var hash2 = sha256HMAC.doFinal(dataString.getBytes());

    var calculatedHash = Hex.toHexString(hash2);

    return calculatedHash.equals(receivedHash);
  }

  private byte[] getSecretHashByInitData() throws GeneralSecurityException {
    var sha256HMAC = Mac.getInstance(HMAC_SHA256);
    var secretKeySpec = new SecretKeySpec(TG_SECRET_KEY.getBytes(), HMAC_SHA256);
    sha256HMAC.init(secretKeySpec);

    return sha256HMAC.doFinal(tgBotToken.getBytes());
  }

  private Map<String, String> parseQueryString(String queryString) {
    return StreamEx.of(queryString.split("&"))
        .map(pair -> pair.split("=", 2))
        .mapToEntry(
            keyValue -> URLDecoder.decode(keyValue[0], UTF_8),
            keyValue -> URLDecoder.decode(
                keyValue.length > 1 ? keyValue[1] : "", UTF_8
            )
        )
        .toMap();
  }
}

File: solo-leveling-gateway-service/src/main/kotlin/com.sleepkqq.sololeveling/gateway/service/TgHashService.kt
package com.sleepkqq.sololeveling.gateway.service

import org.bouncycastle.crypto.digests.SHA256Digest
import org.bouncycastle.crypto.macs.HMac
import org.bouncycastle.crypto.params.KeyParameter
import org.bouncycastle.util.encoders.Hex
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import java.net.URLDecoder
import java.nio.charset.StandardCharsets

@Service
class TgHashService(
	@Value($$"${app.telegram.bot.token}") private val tgBotToken: String
) {

	private companion object {
		const val TG_SECRET_KEY = "WebAppData"

		const val HASH_FIELD = "hash"

		const val QUERY_DELIMITER = "&"
		const val KEY_VALUE_DELIMITER = "="
	}

	fun checkHash(initData: String, hash: String): Boolean {
		return validateHash(parseQuery(initData), hash)
	}

	private fun validateHash(parsedQuery: String, receivedHash: String): Boolean {
		val tgBotTokenHash = getHash(
			TG_SECRET_KEY.toByteArray(StandardCharsets.UTF_8),
			tgBotToken
		)
		val dataHash = getHash(tgBotTokenHash, parsedQuery)

		val calculatedHash = Hex.toHexString(dataHash)
		return calculatedHash == receivedHash
	}

	private fun getHash(keyBytes: ByteArray, data: String): ByteArray {
		val hmac = HMac(SHA256Digest())
		hmac.init(KeyParameter(keyBytes))

		val dataBytes = data.toByteArray(StandardCharsets.UTF_8)
		hmac.update(dataBytes, 0, dataBytes.size)

		val hashBytes = ByteArray(hmac.macSize)
		hmac.doFinal(hashBytes, 0)

		return hashBytes
	}

	private fun parseQuery(queryString: String): String = queryString.split(QUERY_DELIMITER)
		.map { it.split(KEY_VALUE_DELIMITER, limit = 2) }
		.map {
			val key = URLDecoder.decode(it[0], StandardCharsets.UTF_8)
			val value = if (it.size > 1) URLDecoder.decode(it[1], StandardCharsets.UTF_8) else ""
			key to value
		}
		.filterNot { (key, _) -> key == HASH_FIELD }
		.sortedBy { it.first }
		.joinToString("\n") { "${it.first}=${it.second}" }
}